### ISSUE TITLE
clearing /data/meta and regenerating, causes blank struct pages

### ISSUE THREAD

**Issue by TheKojukinator** [cosmocode/dokuwiki-plugin-struct#336]
Hello, I've created a schema to inventory my comics. I've been experimenting a lot and clearing the slate a bunch of times, part of this is clearing out my /data/cache/ and /data/meta/.

I've noticed that when the page.meta file is gone for a page that I have a schema assigned to, with data in it, when I load that page it doesn't show the data structure. I was looking at what page.meta content is generated and noticed something interesting.

So to recreate this, I...

1. Clear all my comics meta from /data/meta/inv/comics/*.*
2. Go to page where all comics are listed, like the Sitemap, and refresh it a bunch of times, this seems to speed up *.meta file generation for all my comics.
3. Go to a particular comics' page to see if my struct data shows up.
4. Notice that only my standard dokuwiki page header is shown.
5. Click edit page, and cancel straight out of it without making any changes.
6. Notice that my header and my struct data is showing as it should finally.

I've observed the contents of the *.meta file and noticed it changes between step 2, 3, and 5.

At step 2, when I visited the Sitemap and let a new *.meta generate for a comic, here is what the contents look like...
`a:2:{s:7:"current";a:8:{s:4:"date";a:2:{s:7:"created";i:1501875681;s:8:"modified";i:1501875681;}s:4:"user";s:0:"";s:7:"creator";s:0:"";s:5:"title";s:37:"Athena Inc. Vol. 1: Manhunter Project";s:11:"description";a:2:{s:15:"tableofcontents";a:1:{i:0;a:4:{s:3:"hid";s:33:"athena_inc_vol_1manhunter_project";s:5:"title";s:37:"Athena Inc. Vol. 1: Manhunter Project";s:4:"type";s:2:"ul";s:5:"level";i:1;}}s:8:"abstract";s:37:"Athena Inc. Vol. 1: Manhunter Project";}s:8:"internal";a:2:{s:5:"cache";b:1;s:3:"toc";b:1;}s:8:"relation";a:1:{s:10:"firstimage";s:0:"";}s:11:"last_change";a:3:{s:4:"date";N;s:4:"user";N;s:2:"ip";N;}}s:10:"persistent";a:3:{s:4:"date";a:1:{s:7:"created";i:1501875681;}s:4:"user";s:0:"";s:7:"creator";s:0:"";}}`

At step 3, when I actually load the page of that comic in dokuwiki and notice that my struct data isn't rendering, the *.meta contents for that file change to...
`a:2:{s:7:"current";a:8:{s:4:"date";a:2:{s:7:"created";i:1501875681;s:8:"modified";i:1501875681;}s:4:"user";s:0:"";s:7:"creator";s:0:"";s:11:"last_change";a:3:{s:4:"date";N;s:4:"user";N;s:2:"ip";N;}s:5:"title";s:37:"Athena Inc. Vol. 1: Manhunter Project";s:11:"description";a:2:{s:15:"tableofcontents";a:1:{i:0;a:4:{s:3:"hid";s:33:"athena_inc_vol_1manhunter_project";s:5:"title";s:37:"Athena Inc. Vol. 1: Manhunter Project";s:4:"type";s:2:"ul";s:5:"level";i:1;}}s:8:"abstract";s:353:"Athena Inc. Vol. 1: Manhunter Project
Comic CollectionCoverTitleAthena Inc. Vol. 1: Manhunter ProjectSeriesAthena Inc.Author(s)Brian HaberlinArtist(s)Jay AnacletoGenre(s)Action, Science FictionPublisherImage ComicsFormat(s)PaperbackPages240Published2003/09/14Goodreads<https://www.goodreads.com/book/show/1317834.Athena_Inc_Volume_1>ISBN-139781582403038";}s:8:"relation";a:2:{s:10:"references";a:1:{s:16:"inv:comics:start";b:1;}s:10:"firstimage";s:58:"https://images.gr-assets.com/books/1347840493l/1317834.jpg";}s:8:"internal";a:2:{s:5:"cache";b:1;s:3:"toc";b:1;}}s:10:"persistent";a:4:{s:4:"date";a:1:{s:7:"created";i:1501875681;}s:4:"user";s:0:"";s:7:"creator";s:0:"";s:11:"last_change";b:0;}}`

Finally, at step 5, after going to edit and cancelling out, the *.meta changes to...
`a:2:{s:7:"current";a:8:{s:4:"date";a:2:{s:7:"created";i:1501875681;s:8:"modified";i:1501875681;}s:4:"user";s:0:"";s:7:"creator";s:0:"";s:11:"last_change";b:0;s:5:"title";s:37:"Athena Inc. Vol. 1: Manhunter Project";s:11:"description";a:2:{s:15:"tableofcontents";a:1:{i:0;a:4:{s:3:"hid";s:33:"athena_inc_vol_1manhunter_project";s:5:"title";s:37:"Athena Inc. Vol. 1: Manhunter Project";s:4:"type";s:2:"ul";s:5:"level";i:1;}}s:8:"abstract";s:37:"Athena Inc. Vol. 1: Manhunter Project";}s:8:"internal";a:2:{s:5:"cache";b:1;s:3:"toc";b:1;}s:8:"relation";a:1:{s:10:"firstimage";s:0:"";}}s:10:"persistent";a:4:{s:4:"date";a:1:{s:7:"created";i:1501875681;}s:4:"user";s:0:"";s:7:"creator";s:0:"";s:11:"last_change";b:0;}}`

It's becoming quite troublesome to have to open each page that I have struct data on, just so I can click edit and cancel to make it render properly. I suspect, whatever mechanism is generating fresh *.meta for pages, isn't handling the associated struct data properly.

Let me know if there's anything else I can provide to assist with troubleshooting. In case it might help, I attached the rendered HTML for the comic page in my example, for comparison between steps 2 and 5.

Thanks,
Steve

[comics_broken_meta_step3.txt](https://github.com/cosmocode/dokuwiki-plugin-struct/files/1316512/comics_broken_meta_step3.txt)
[comics_broken_meta_step5.txt](https://github.com/cosmocode/dokuwiki-plugin-struct/files/1316511/comics_broken_meta_step5.txt)

**Comment by TheKojukinator** [comment]
I just figured out that appending `&purge=1` to the URL of the affected page, will force it to render properly regardless of cache/metadata presence. My issue now of course is that I need to do this for every page on my wiki with struct data ~_~

**Comment by micgro42** [comment]
> part of this is clearing out my /data/cache/ and /data/meta/.

... Don't.

What is the problem you are trying to solve with this anyway? o.O

You can try saving the wiki config. This should globally invalidate the cache and may help with your problem.

**Comment by TheKojukinator** [comment]
I've moved my pages around a lot and re-organized namespaces while settling on how my wiki will be structured. After all this experimentation I had a bunch of leftover stuff in Meta that I figured wasn't necessary, so I just wiped it all out and figured it would regenerate as it has before I used the struct plugin. 

I wasn't deleting all of Meta, I know the sqlite db is in there, so I kept that around.

I understand that in normal use cases this shouldn't be necessary. But I can imagine a scenario where I only backup the Pages, and the sqlite db from Meta, and then restore those to a fresh Dokuwiki deployment. After performing the schema assignment, I would then visit the pages with struct data, and they would be blank, until I attempt to edit/cancel each one, or visit them with `&purge=1`.

I wrote a script the enumerates all my Pages, generates URLs with `&purge=1` appended, and visits them, in case I decide to bomb my Meta again.

In the absence of page data in Meta, if the plugin could generate it properly without having to use `&purge=1`, that would basically solve this fringe issue. But I don't know if that's possible, as I don't fully understand meta generation.
